File: exercices/exercice1-2-5/binaryTree.js
class Noeud {
    constructor(val) {
        this.valeur = val;
        this.gauche = undefined;
        this.droite = undefined;
        this.parent = undefined;
    }

    // Affiche la valeur du noeud et la valeur de ses deux enfants et de son parent
    toString() {
        var out = "Noeud " + this.valeur + ":  L";

        this.gauche === undefined ? out += "-" : out += this.gauche.valeur;
        out += " R";

        this.droite === undefined ? out += "-" : out += this.droite.valeur;
        out += " P";

        this.parent === undefined ? out += "-" : out += this.parent.valeur;
        console.log(out);
    }
}
class Arbre {
    constructor() {
        this.racine = undefined;
    }


    trouverNoeudArbre(node, valeur) {
        if (node != undefined) {
            if(node.valeur === valeur) {
                return node;
            } else if (node.valeur > valeur) {
                return this.trouverNoeudArbre(node.gauche, valeur);
            } else if (node.valeur < valeur) {
                return this.trouverNoeudArbre(node.droite, valeur);
            } 
        }
        return undefined;
    }
    //Méthode pour trouver une valeur donnée dans un arbre binaire de recherche
    trouverNoeud(valeur) {
        if(this.racine === undefined)
            return false;
        return this.trouverNoeudArbre(this.racine, valeur);
    }

    ajouterNoeudArbre(noeud, valeur) {
        // console.log("ajouterNoeud(node, valeur) ");
        if (noeud !== undefined) {
            if (valeur < noeud.valeur) {
                if(noeud.gauche === undefined) {
                    noeud.gauche = new Noeud(valeur);
                    noeud.gauche.parent = noeud;
                } else {
                    this.ajouterNoeudArbre(noeud.gauche, valeur);
                }
            }
            else if (valeur > noeud.valeur) {
                if(noeud.droite === undefined) {
                    noeud.droite = new Noeud(valeur);
                    noeud.droite.parent = noeud;
                } else {
                    this.ajouterNoeudArbre(noeud.droite, valeur);
                }
            } else {
                console.log("Node already exists");
            }
        }
    }

    ajouterNoeud(valeur) {
        if (this.racine === undefined) {
            this.racine = new Noeud(valeur);
        } else {
            this.ajouterNoeudArbre(this.racine, valeur)
        }
    }

    maxValue(noeud) {
        if (noeud.droite === undefined)
              return noeud.valeur;
        else
              return this.maxValue(noeud.droite);
    }

    minValue(noeud) {
        if (noeud.gauche=== undefined)
              return noeud.valeur;
        else
              return this.minValue(noeud.gauche);
    }
    //Méthode pour supprimer un noeud
    supprimerNoeudArbre(node, valeur) {
        if (valeur < node.valeur) {
            if (node.gauche !== undefined)
                  return this.supprimerNoeudArbre(node.gauche, valeur);
            else
                  return false;
        } else if (valeur > node.valeur) {
            if (node.droite !== undefined)
                  return this.supprimerNoeudArbre(node.droite, valeur);
            else
                  return false;
        } else {
                if (node.gauche !== undefined && node.droite !== undefined) {
                    node.valeurGauche = this.minValue(node.gauche);
                    node.valeurDroite = this.maxValue(node.droite);

                    if (node.valeur - node.valeurGauche < node.valeurDroite - node.valeur) {
                        this.supprimerNoeudArbre(node.gauche, node.valeurGauche);
                        node.valeur = node.valeurGauche;
                    } else {
                        this.supprimerNoeudArbre(node.droite, node.valeurDroite);
                         node.valeur = node.valeurDroite;
                    }

                } else if (node.parent.gauche === node) {
                    node.parent.gauche = (node.gauche !== undefined) ? node.gauche : node.droite;
                } else if (node.parent.droite === node) {
                    node.parent.droite = (node.gauche !== undefined) ? node.gauche : node.droite;
                }
                return true;
        }
    }

    supprimerNoeud(valeur) {
        if(this.racine.valeur === valeur) {
            this.racine = undefined;
        } else {
            this.supprimerNoeudArbre(this.racine, valeur);
        }

    }


    //Méthode pour afficher l’arbre selon un parcours infixe
    //Cette méthode doit retournée un tableau contenant la valeur des noeuds
    infixeNoeud(noeud, array) {
        if(noeud !== undefined) {
            if(noeud.gauche !== undefined) {
                this.infixeNoeud(noeud.gauche, array);
            }
            if(noeud.valeur !== undefined) {
                array.push(noeud.valeur);
            }
            if(noeud.droite !== undefined) {
                this.infixeNoeud(noeud.droite, array);
            }
        }
    }

    infixe() {
        var array = [];
        if(this !== undefined) {
            if(this.racine !== undefined) {
                this.infixeNoeud(this.racine, array);
                return array;
            }
        }

    }

    //Méthode pour afficher la valeur d'un noeud à partir de sa valeur
    printNoeud (valeur) {
        let noeud = this.trouverNoeud(valeur);
        if (noeud !== undefined) noeud.toString();
    }
}

File: exercices/exercice1-2-5/binaryTree.py
class Noeud:
    def __init__(self, v, parent=None):
        self.valeur = v
        self.parent = parent
        self.gauche = None
        self.droite = None

    def toString(self):
        out = "Noeud " + str(self.valeur) + ":  L";
        if (self.gauche is None):
            out += "-"
        else:
            out += str(self.gauche.valeur)
        out += " R";
        if (self.droite is None):
            out += "-"
        else:
            out += str(self.droite.valeur)
        out += " P";
        if (self.parent is None):
            out += "-"
        else:
            out += str(self.parent.valeur)

        print(out)

    def trouverNoeud(self, valeur):
        if self is not None and valeur is not None:
            if self.valeur == valeur:
                return self;
            elif self.valeur > valeur and self.gauche is not None:
                return self.gauche.trouverNoeud(valeur);
            elif self.valeur < valeur and self.droite is not None:
                return self.droite.trouverNoeud(valeur);
        return None;

    def minValue(self):
        if self.gauche is None:
            return self.valeur;
        else:
            return self.gauche.minValue();

    def maxValue(self):
        if self.droite is None:
            return self.valeur;
        else:
            return self.droite.maxValue();

    def ajouterNoeud(self, valeur):
        if self is not None:
            if valeur < self.valeur:
                if self.gauche is None:
                    self.gauche = Noeud(valeur);
                    self.gauche.parent = self;
                else:
                    self.gauche.ajouterNoeud(valeur);


            elif valeur > self.valeur:
                if self.droite is None:
                    self.droite = Noeud(valeur);
                    self.droite.parent = self;
                else:
                    self.droite.ajouterNoeud(valeur);

            else:
                print("Node already exists");

    # Méthode pour supprimer un noeud
    def supprimerNoeud(self, valeur):
        if valeur is None:
            return;
        if valeur < self.valeur:
            if self.gauche is not None:
                return self.gauche.supprimerNoeud(valeur);
            else:
                return False;
        elif valeur > self.valeur:
            if self.droite is not None:
                return self.droite.supprimerNoeud(valeur);
            else:
                return False;
        else:
            if self.gauche is not None and self.droite is not None:
                valeurGauche = self.gauche.minValue();
                valeurDroite = self.droite.maxValue();

                if self.valeur - valeurGauche < valeurDroite - self.valeur:
                    self.gauche.supprimerNoeud(valeurGauche);
                    self.valeur = valeurGauche;
                else:
                    self.droite.supprimerNoeud(valeurDroite);
                    self.valeur = valeurDroite;


            elif self.parent.gauche == self:
                if self.gauche is not None:
                    self.parent.gauche = self.gauche;
                else:
                    self.parent.gauche = self.droite;
            elif self.parent.droite == self:
                if self.gauche is not None:
                    self.parent.droite = self.gauche;
                else:
                    self.parent.droite = self.droite;

            return True;

    # Méthode pour afficher l’arbre selon un parcours infixe
    # Cette méthode doit retournée un tableau contenant la valeur des noeuds
    def infixeNoeud(self, array):
        if self is not None:
            if self.gauche is not None:
                self.gauche.infixeNoeud(array);

            if self.valeur is not None:
                array.append(str(self.valeur));

            if self.droite is not None:
                self.droite.infixeNoeud(array);


class Arbre:
    def __init__(self):
        self.racine = None

    # Méthode pour trouver une valeur donnée dans un arbre binaire de recherche
    def trouverNoeud(self, val):
        if self.racine is None:
            return False;
        return self.racine.trouverNoeud(val);

    # Méthode pour ajouter un noeud
    def ajouterNoeud(self, val):
        if self.racine is None:
            self.racine = Noeud(val);
        else:
            self.racine.ajouterNoeud(val)

    # Méthode pour supprimer un noeud
    def supprimerNoeud(self, val):
        if self.racine.valeur == val:
            self.racine = None;
        else:
            self.racine.supprimerNoeud(val);

    # Méthode pour afficher l’arbre selon un parcours infixe
    # Cette méthode doit retournée un tableau contenant la valeur des noeuds
    def infixe(self):
        array = [];
        if self is not None:
            if self.racine is not None:
                self.racine.infixeNoeud(array);
                return array;

    # Méthode pour afficher la valeur d'un noeud à partir de sa valeur
    def printNoeud(self, val):
        noeud = self.trouverNoeud(val)
        if noeud is not None:
            noeud.toString()